Comments option

olleeriksson opened the pull request:
I am just creating this draft PR for you to be able to give your opinion before I clean it up. So I will discard it and put up an updated one when I'm done. It has not language files, and some commented code, and I haven't tested it fully, but if you want to try it feel free to go ahead.

I called the option "Comments: only played move" and "Comments: move + responses". Thoughts?

As you can see below I managed to add a bit of space between comments. I think that make it much easier to tell them apart. I also moved the comment from the move that was just played to the top, and the comments from the coming available responses from the player below. I also made the text size of the comments 0.9em instead of 1.0em. The underlines move names are taken from the code programmatically. It only shows on comments from the responses you have available. The rest is text from the comment itself:

![image](https://user-images.githubusercontent.com/41130048/203454821-de5eb08a-cdde-4970-b835-96e96e1a02c7.png)

Here is another example where the user is showing only comments from the move that has been played.

![image](https://user-images.githubusercontent.com/41130048/203454911-8e189666-2f10-4e50-9008-52798442acdf.png)

And here is one with even longer text:

![image](https://user-images.githubusercontent.com/41130048/203455007-3458dc74-4db0-42df-b3c3-0d6e82588380.png)

It also occurred to me that these two icons (marked with an arrow) could be used if one wants to have the icon flip between two different states when changing the comments option:

![image](https://user-images.githubusercontent.com/41130048/203455070-5beac2f0-5db9-4c05-8415-20413da62811.png)

olleeriksson committed:
Opposite side comments option

olleeriksson commented:
If it helps: here is a study with a few moves from the Sicilian opening with some comments:
https://lichess.org/study/InuSjp1l

And the same repertoire imported into LiStudy:
https://listudy.org/en/studies/4l9p9e-sicilians-with-comments
https://listudy.org/en/studies/oahxer-sicilians-with-comments-black

I've started the comment on every move with "White|Black plays xxxx" so it's easy to tell where each comment comes from.

olleeriksson committed:
Added a link to analyze the position with Lichess Analysis Board (#134)

ArneVogel committed:
add speech bubble icons

ArneVogel commented:
I added the icons.

Some thoughts:

1. I think having the "White|Black plays xxxx - " added automatically would be nice.
2. Since we already know the move for which we display the comment, how about making the move interact able:  When you hover (or click on mobile) show an arrow showing the move on the board.
3. Going even further, do the same with the moves in the actual comment. If it is a legal move in the current position make it interact able as in point 2.

olleeriksson commented:
Alright, I'll see what I can do..

olleeriksson committed:
Refactored the displaying of comments a little

olleeriksson committed:
Added language files changes for comments

olleeriksson committed:
Re-arranged the options a little

olleeriksson commented:
Ok, so I tried to do what you suggested, and I refactored the code a little to prepare for adding things like being able to click the move to make it, but I couldn't get the onclick handler to work because of browser security issues. I think maybe the most useful thing would be highlighting the move with an arrow, but I didn't want to start implementing that in the middle of everything. I think I have to limit myself now. I've spent so many late nights that I need to wrap it up a bit. But it should be an interesting future thing to add. 

So I left it like this, and the PR is ready for review now. I did listen to your comment number 1 with a slight modification. Here is a screenshot of how it looks now. I put in a newline between the move and the comment, and added "White|Black resonse [Move]". I skipped the word "response" or "played" for the move that was just played, because I felt it helped make it stand out a little. Also, if you're only showing comments from the played move, I am not even adding the color "Black|White". I felt it wasn't needed then.

If you still want me to use the wording that you suggested or make any changes of course I'll do that. But I will leave highlighting the arrow and being able to click on the move for a later time.

Oh and BTW, I already have the "Options in localstorage" code ready, but I'll wait to upload it until this one has gone through in case there are any code conflicts.

Anyway, a sample:

![image](https://user-images.githubusercontent.com/41130048/203670722-f2230b77-0e19-40c4-b375-b833a7b047f7.png)

Showing comments only from the played move:

![image](https://user-images.githubusercontent.com/41130048/203670823-a4c38eb8-4d9e-4e2f-be26-4ea67767df73.png)

olleeriksson committed:
Merge branch 'master' into feature/opposite-side-comments-option

olleeriksson committed:
Language files updates.

olleeriksson commented:
It's ready now.

olleeriksson commented:
Oh and I don't know how you prefer to merge things, but there are a couple of commits this time, so if you prefer the straight line type of history it might be an idea to do a squash merge. Anyway, just so you're aware of it.

ArneVogel commented on assets/js/study.js:
```diff
@@ -173,27 +175,79 @@ function give_hints(access, once) {
-    // give comment on the current move if there are move to give hint to
-    if (to_hint.length > 0 && cm.comments != undefined && cm.comments[0] != undefined && cm.comments[0].text != undefined) {
-        comment += cm.comments[0].text.trim();
+    // Get current move if it has a comment
```
The comments should only be displayed when a arrow hint is to be displayed. Otherwise its just a spoiler.

E.g. this move was played enough times so no hint is given, but the comment still spoils the right move.
![image](https://user-images.githubusercontent.com/26939450/204079601-1c475a23-06b5-4e3e-a4d9-3f4128f88f68.png)

olleeriksson replied:
Well..  I had actually intentionally changed that. But yeah I see what you mean. I think my thinking was that even if a comment is visible, it only provides some background on the position, maybe describes the name of the position, or has a longer description like in your screenshot above. It would spoil the move if you read the comment, you're right. But my thinking is that if you read the text is has to be a conscious decision, so it should be easy enough to avoid it if you don't want the help. An arrow that shows up when you don't want to see it is a lot harder to "unsee", so that's why I personally think it's much more important to make sure the user doesn't see the arrow hints than the comment hints. But, you're right, it might give a hint to the user and spoil the move if the comment is easy enough to read and interpret when you don't want to. 

 I'll change it back if you want to, no problem. Let me know what you decide!

ArneVogel replied:
I would prefer to revert it back.

But the idea gave me another thought: maybe to create a new review mode.

1. Unconditionally: Show all arrows
2. Unconditionally: Show all comments
3. Show the PGN on the left
3.1. Allow the moves in the pgn to be clickable like on lichess to go to the move.
4. With the addition of statistics collection
4.1. Highlight "weak" moves, where the most errors were made in training mode in the pgn
4.2. When on the "weak" move on the board somehow make it obvious that more attention should be given to this move

Of course out of scope for this pr but just commenting as a idea for the future.

olleeriksson replied:
Hmm.. It gets complicated when you think about it. The comments for the played move is usually the opponents moves, unless you have a slow move delay I suppose, in case I think your own played response's comment would be "the played move" (haven't tested that to be honest, and I'm out hiking today :)). But also, would people really read their own response's comment in the second or so after the response was played? I don't think so. 

So, if we connect the displaying of arrows to the displaying of comments, then you won't see the opponents (the played move's) comments, because there are never any arrows shown for the opponent's moves. I guess that's how it was before though. But it's a little strange I think. Turning off your own arrows turns of the opponents comments. 

How about:
Comments: played move (always).           (Ie the opponents move in practicality)
Comments: all moves (always)
Comments: played move always + responses with arrows
Comments: responses with arrows

The problem I have is finding a short enough text string to describe these modes.

Or, another option altogether:

We **always** display comments for the opponent's moves (not only when the played move is his). And we allow the option to control if **your** move's comments are shown or not:

Comments: all your moves
Comments: for arrowed moves

But maybe just with better wording.

Or leave it as is in the PR, where the comments (all comments) are controlled independently of the arrows.

I have trouble deciding what is best here. ;)

olleeriksson replied:
Wait a minute.... there's another aspect to this altogether. 

Maybe, just maybe, apart from comments on the move the opponent just made, maybe the user **doesn't** want to see comments for possible upcoming responses, but comments on the **one single** move the user played one move back in time. I mean it's only **that** comment that describes the position we ended up with. The problem was in most cases the opponent (ai) then played his move instantly and without delay, so we never had a chance to read the comment of the position that we gave ourselves. 

We probably should think about comments more in terms of a position. You want the comments on the current position. But do you also want them on the previous position (after the move you played), and do you want them on all possible upcoming positions depending on how many candidate moves you have to choose from.

Correct me if I am wrong, but I don't think that is how the code worked before. So making such a change would break backwards compatibility.

Man it gets complicated. :)

ArneVogel replied:
> maybe the user doesn't want to see comments for possible upcoming responses

I don't get this. I think it would be weird to be one move "behind". 

From https://lichess.org/study/Jn2JyhRT/6nJ8VQoA :

From whites perspective currently it is what I would expect (and what is implemented) while arrows are shown:
![image](https://user-images.githubusercontent.com/26939450/204099254-3d406302-d8fc-47d5-b579-1b74b764b2bc.png)

I think this should be the default while arrows are shown, with no arrows in the position only the opponent moves should be shown.
And the comment option should then be on if comments are shown in this way, or no comments are shown at all.

olleeriksson replied:
Yeah after thinking more about it, I think you're definitely right!

I think the key here is whether most comments talks mostly about the current position or any follow-up moves the other side may have. And it's probably more likely that a comment focuses on the current position than future responses, so for that reason it's better to show current and future comments.

So I like your ideas and will update the PR accordingly.

But... to achieve what you want of not spoiling any information, I think we need to hide **all** comments, including the "played move". Consider this simple example. If for example the comment on 1.e4 says "Black plays the most common opening, and white can respond with e5 or c5" then that definitely spoils the position for black.

Maybe the whole idea of having an option for controlling whether comments from one or both sides are displayed is a bit of overkill. Especially now that it's clearer visually what move a comment belongs to.

So how about this: I fix the other issues you mentioned below. And we also keep the slightly more refined visual changes in how comments are displayed. And then make the option as follows instead:

Comments: when arrows show
Comments: always on
Comments: always off

Personally I will practice with arrows always off, and turn them on using the Hints link, but I always want the comments showing. And the suggestion above allows me to get that.

olleeriksson replied:
I uploaded a new commit to modify these things.

So: 

- Removed the option to choose between "Comments: played move" and "Comments: played move + responses".
- Replaced with option to choose between "Comments: when arrows show", "Comments: always on", and "Comments: hidden".
- Not showing comments entered before the first move in the PGN. I think that should be a reasonable solution.
- Not calling the opening moves "responses".

ArneVogel commented on lib/listudy_web/templates/study/show.html.eex:
```diff
@@ -146,7 +145,10 @@
   i18n.fast = "<%= dgettext "study", "fast" %>";
   i18n.medium = "<%= dgettext "study", "medium" %>";
   i18n.instant = "<%= dgettext "study", "instant" %>";
+  i18n.comments_only_move = "<%= dgettext "study", "Comments: only played move" %>";
```
Maybe "only my moves"?

olleeriksson replied:
I had written "my moves" first, before I realized it's not a comment on **your** move, it's actually the **opponent's** move. It's the responses that are yours. "The played move" or "the last move" or something was the best I could come up with.

ArneVogel commented on lib/listudy_web/templates/study/show.html.eex:
```diff
@@ -146,7 +145,10 @@
   i18n.medium = "<%= dgettext "study", "medium" %>";
   i18n.instant = "<%= dgettext "study", "instant" %>";
+  i18n.comments_only_move = "<%= dgettext "study", "Comments: only played move" %>";
+  i18n.comments_move_and_responses = "<%= dgettext "study", "Comments: move + responses" %>";
```
Adjusting with above maybe: "my moves + responses".

olleeriksson replied:
Same as above.

ArneVogel reviewed (commented):
Some nits

ArneVogel commented:
Something I just noticed: there is currently no indication if a comment is cut off:
No indication of scroll:
![image](https://user-images.githubusercontent.com/26939450/204082043-a5be5e11-cbd0-4c35-b45e-e04dbc2e9533.png)

When there is text below:
![image](https://user-images.githubusercontent.com/26939450/204082054-e2b171c8-d91b-45a3-b750-1ac59c6d2034.png)

Just noting for further improvements. After a short investigation I am not sure if its possible just using css so this would probably have to get solved in js.

olleeriksson commented:
Yeah I noticed that large comments don't behave great, but I didn't want to
start making too many modifications either. But something should be done to
handle large amounts of text in the comments.

On Sat, Nov 26, 2022, 10:38 ArneVogel ***@***.***> wrote:

> Something I just noticed: there is currently no indication if a comment is
> cut off:
> No indication of scroll:
> [image: image]
> <https://user-images.githubusercontent.com/26939450/204082043-a5be5e11-cbd0-4c35-b45e-e04dbc2e9533.png>
>
> When there is text below:
> [image: image]
> <https://user-images.githubusercontent.com/26939450/204082054-e2b171c8-d91b-45a3-b750-1ac59c6d2034.png>
>
> Just noting for further improvements. After a short investigation I am not
> sure if its possible just using css so this would probably have to get
> solved in js.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ArneVogel/listudy/pull/135#issuecomment-1328014610>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJZZQQHWAO4J732NWOSUMNTWKHK7RANCNFSM6AAAAAASIOC65Y>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

ArneVogel commented:
Some issues I found while exploring comment options (from https://lichess.org/study/Jn2JyhRT/6nJ8VQoA)

Comments unattached to any more before the game bugs out, and opening moves should probably not be a response
![image](https://user-images.githubusercontent.com/26939450/204099134-1c97ae0f-7d73-445e-9491-237848a42267.png)

olleeriksson committed:
Changed the comment option according to review comments

No longer choosing between "played move" and "played move + responses".
Now selecting between "when arrows show", "always on", and "hidden".
Also not showing comments before the first move.
And not calling the opening moves "responses".

ArneVogel commented:
Could you please add the comment that would be placed before the study back? As with the comment above just without the "Black NaN...undefined:". Otherwise looks good to me.

olleeriksson committed:
Displaying comments before the first move

olleeriksson commented:
Alright. There you go!

ArneVogel commented:
Thanks a lot!
